File: match.rb
=begin
# This is an example data structure that will be passed into similarity:
one = {
	:freqs => {
		"sea"=> 1,
		"salt"=> 1,
		"I"=> 1,
		"the"=> 1,
		"love"=> 1,
		"chocolate"=> 1
	},
	:nonstoplist_freqs => {
		"lol" => 1,
		"omg" => 2
	},
	:avg_len => 0,
	:total_words => 0
}



similarity(one, two) returns a result object

result = {
	:score => 0.0 #(0 to 1)
	:matchUser => userName
}
=end

def similarity(one, two)

	score = 0.0
        score2 = 0.0

	words = one[:freqs].keys
	words2 = two[:freqs].keys

	words.each do |word, value|
		if two[:freqs].has_key? word
			score += [value, two[:freqs][word]].min
		end
	end

        #normalize unique word count by dividing by the total number of words 
        unique1 = words.length().to_f / one[:total_words]
        unique2 = words.length().to_f / two[:total_words]

        score2 = (unique1 - unique2) ** 2

	# normalize score by higher total_words
	# if we don't do this, people who talk A LOT will
	# tend to have higher scores
	score /= [one[:total_words], two[:total_words]].max


	# penalize pairs with different average sentence lengths
	smaller = [one[:avg_len], two[:avg_len]].min
	larger = [one[:avg_len], two[:avg_len]].max

	score *= smaller/larger
        
        #now penalize pairs with differing average unique number of words
	smaller = [unique1, unique2].min
	larger = [unique1, unique2].max

	score *= smaller/larger
        
	return score
end

File: match.py
one = {
	freqs: {
		"sea": 1,
		"salt": 1,
		"I": 1,
		"the": 1,
		"love": 1,
		"chocolate": 1
	},
	avg_len: 0,
	total_words: 0
}

two = {
	freqs: {
		"sea": 1,
		"salt": 1,
		"I": 1,
		"the": 1,
		"love": 1,
		"chocolate": 1
	},
	avg_len: 0,
	total_words: 0
}

class User:
	def __init__(self, freqs, avg_len):
		self.freqs = freqs
		self.avg_len = avg_len

	def get_total_words(self):
		return sum(self.freqs.values())

	def get_words(self):
		return self.freqs.keys()

def similarity(one, two):
"""
	Given the statistics for two people, calculate their similarity metric.
	Inputs ('one', 'two') are objects 
]
"""
	score = 0.0
	words = one['freqs'].keys()
	for word in words:
		if word in two['freqs']:
			score += min(one['freqs'][word], two['freqs'][word])

	# normalize score by higher total_words
	# if we don't do this, people who talk A LOT will
	# tend to have higher scores

	score /= max(one['total_words'], two['total_words'])

	# reward pairs with similar lengths
	small = min(one['avg_len'], two['avg_len'])
	large = max(one['avg_len'], two['avg_len'])
	score *= min/max

	return score